contracts/FeeDisbursal/FeeDisbursal.sol

```diff
@@ -0,0 +1,378 @@
+    {    
+        // cycle through all investors collect the claimable and recycleable funds
+        uint256 totalAmount = 0;
+        for (uint256 i = 0; i < investors.length; i += 1) {
+            (uint256 claimableAmount,) = claimablePrivate(token, investors[i], until, true);
+            totalAmount += claimableAmount;
+        }
+        return totalAmount;
+    }
+
+    /// @notice get current controller
+    function feeDisbursalController()
+        public
+        constant
+        returns (IFeeDisbursalController)
+    {
+        return _feeDisbursalController;
+    }
+
+    /// @notice update current controller
+    function changeFeeDisbursalController(IFeeDisbursalController newController)
+        public
+    {
+        require(_feeDisbursalController.onChangeFeeDisbursalController(msg.sender, newController));
+        address oldController = address(_feeDisbursalController);
+        _feeDisbursalController = newController;
+        emit LogChangeFeeDisbursalController(oldController, address(newController), msg.sender);
+    }
+
+    /// @notice implementation of tokenfallback, calls the internal disburse function
+    /// legacy onTokenTransfer is also supported via imported file
+    function tokenFallback(address wallet, uint256 amount, bytes data)
+        public
+    {
+        // cast and check pro rata token
+        // fallback is neumark
+        BasicSnapshotToken proRataToken;
+        if (data.length != 0) 
+           proRataToken = BasicSnapshotToken(decodeAddress(data));
+        else
+            proRataToken = UNIVERSE.neumark();
+        disburse(msg.sender, wallet, amount, proRataToken);
+    }
+
+
+    ////////////////////////
+    // Private functions
+    ////////////////////////
+
+    /// @notice create a new disbursal
+    /// @param token address of the token to disburse
+    /// @param disburser address of the actor disbursing (e.g. eto commitment)
+    /// @param amount amount of the disbursable tokens
+    /// @param proRataToken address of the token that defines the pro rata
+    function disburse(address token, address disburser, uint256 amount, BasicSnapshotToken proRataToken)
+    internal
+    {
+        require(_feeDisbursalController.onDisburse(token, disburser, amount, address(proRataToken)), "");
+
+        uint256 snapshotId = proRataToken.currentSnapshotId();
+        uint256 proRataTokenTotalSupply = proRataToken.totalSupplyAt(snapshotId);
+        require(proRataTokenTotalSupply > 0, "");
+
+        Disbursal[] storage disbursals = _disbursals[token];
+
+        // try to merge with an existing disbursal
+        bool merged = false;
+        for ( uint256 i = disbursals.length - 1; i < disbursals.length; i-- ) {
+            // we can only merge if we have the same snapshot id
+            // we can break here, as continuing down the loop the snapshot ids will decrease
+            Disbursal storage disbursal = disbursals[i];
+            if ( disbursal.snapshotId < snapshotId )
```

Comment:
That does not make sense to me. If we merge disbursals that have different pro rata tokens set, this will end in total chaos. I can't think of the case where this would work, but maybe we can talk about it to be clear.

Reply:
I do not agree. Let's discuss this